src/collision/narrow_phase.rs

```diff
@@ -162,139 +183,457 @@ pub enum NarrowPhaseSet {
     /// add custom systems to this set, or to [`PostProcessCollisions`].
     PostProcess,
+    /// Generates [`ContactConstraint`]s and adds them to [`ContactConstraints`].
+    GenerateConstraints,
     /// Runs at the end of the narrow phase. Empty by default.
     Last,
 }
 
-/// Computes contacts based on [`BroadCollisionPairs`] and adds them to [`Collisions`].
-#[allow(clippy::too_many_arguments)]
-#[allow(clippy::type_complexity)]
-pub fn collect_collisions<C: AnyCollider>(
-    query: Query<(
-        Ref<Position>,
-        Option<&AccumulatedTranslation>,
-        Ref<Rotation>,
-        &C,
-    )>,
+fn collect_collisions<C: AnyCollider>(
+    mut narrow_phase: NarrowPhase<C>,
     broad_collision_pairs: Res<BroadCollisionPairs>,
-    mut collisions: ResMut<Collisions>,
-    narrow_phase_config: Res<NarrowPhaseConfig>,
+    solver_config: Res<SolverConfig>,
+    time: Res<Time>,
 ) {
-    if query.is_empty() {
-        return;
+    let warm_start = solver_config.warm_start_coefficient > 0.0;
+
+    narrow_phase.update(
+        &broad_collision_pairs,
+        warm_start,
+        time.delta_seconds_adjusted(),
+    );
+}
+
+// TODO: It'd be nice to generate the constraint in the same parallel loop as `collect_collisions`
+//       to avoid the extra iteration and queries. This is possible, but it wouldn't work with the current
+//       `PostProcessCollisions` setup.
+fn generate_constraints<C: AnyCollider>(
+    narrow_phase: NarrowPhase<C>,
+    mut constraints: ResMut<ContactConstraints>,
+    solver_config: Res<SolverConfig>,
+    contact_softness: Res<ContactSoftnessCoefficients>,
+    time: Res<Time>,
+) {
+    let warm_start = solver_config.warm_start_coefficient > 0.0;
+    let delta_secs = time.delta_seconds_adjusted();
+
+    // Clear contact constraints.
+    constraints.clear();
+
+    // TODO: Parallelize.
+    for contacts in narrow_phase.collisions.get_internal().values() {
+        let Ok([collider1, collider2]) = narrow_phase
+            .collider_query
+            .get_many([contacts.entity1, contacts.entity2])
+        else {
+            continue;
+        };
+
+        let body1_bundle = collider1
+            .parent
+            .and_then(|p| narrow_phase.body_query.get(p.get()).ok());
+        let body2_bundle = collider2
+            .parent
+            .and_then(|p| narrow_phase.body_query.get(p.get()).ok());
+        if let (Some(body1), Some(body2)) = (
+            body1_bundle.map(|(body, _)| body),
+            body2_bundle.map(|(body, _)| body),
+        ) {
+            // At least one of the bodies must be dynamic for contact constraints
+            // to be generated.
+            if !body1.rb.is_dynamic() && !body2.rb.is_dynamic() {
+                continue;
+            }
+
+            // Generate contact constraints for the computed contacts
+            // and add them to `constraints`.
+            narrow_phase.generate_constraints(
```

Comment:
Can we put this into a struct so that the call becomes cleaner?

Reply:
Not sure if it's worth it unless it could be reused elsewhere. I'm also not sure what the struct could be called